joerd/source/ned.py

```diff
@@ -0,0 +1,183 @@
+        raise
+
+
+def _parallel(func, iterable, num_threads=None):
+    p = Pool(processes=num_threads)
+    threads = []
+
+    for x in iterable:
+        p.apply_async(func, x)
+
+    p.close()
+    return_values = []
+    for t in threads:
+        return_values.append(t.get())
+
+    p.join()
+    return return_values
+
+
+NED_IMG_PATTERN = re.compile('^ned19_' \
+                             '([ns])([0-9]{2})x([0257][05])_' \
+                             '([ew])([0-9]{3})x([0257][05])_' \
+                             '[a-z]{2}_[a-z]+_topobathy_20[0-9]{2}.(zip|img)$')
+
+
+def _ned_parse_filename(fname):
+    m = NED_IMG_PATTERN.match(fname)
+
+    if m:
+        y = int(m.group(2)) + float(m.group(3)) / 100.0
+        x = int(m.group(5)) + float(m.group(6)) / 100.0
+        if m.group(1) == 's':
+            y = -y
+        if m.group(4) == 'w':
+            x = -x
+        return BoundingBox(x, y - 0.25, x + 0.25, y)
+
+    return None
+
+
+def _list_ned_files():
+    ftp = FTP(NED_FTP_SERVER)
+    files = []
+
+    def _callback(zname):
+        bbox = _ned_parse_filename(zname)
+        if bbox is not None:
+            fname = zname.replace(".zip", ".img")
+            files.append((bbox, fname, zname))
+
+    ftp.login()
+    ftp.cwd(NED_BASE_PATH)
+    try:
+        ftp.set_pasv(True)
+        ftp.retrlines('NLST', _callback)
+        ftp.quit()
+    except EOFError:
+        pass
+
+    return files
+
+
+class NED:
+
+    def __init__(self, regions, base_dir='ned', num_download_threads=None):
+        self.regions = regions
+        self.num_download_threads = num_download_threads
+        self.base_dir = base_dir
+
+    def download(self):
+        logger = logging.getLogger('ned')
+        logger.info('Fetching NED index...')
+
+        files = []
+        for bbox, fname, zname in _list_ned_files():
+            if self._intersects(bbox):
+                files.append((fname, zname))
+
+        if not os.path.isdir(self.base_dir):
+            os.makedirs(self.base_dir)
+
+        logger.info("Starting download of %d NED files." % len(files))
+        files = _parallel(_download_ned_file,
+                          [(f, z, self.base_dir) for f, z in files],
+                          num_threads=self.num_download_threads)
+
+        # sanity check
+        for f, z in files:
+            assert os.path.isfile(os.path.join(self.base_dir, f))
+
+        logger.info("Download complete.")
+
+    def buildvrt(self):
+        logger = logging.getLogger('ned')
+        logger.info("Creating VRT.")
+
+        files = []
+        for f in glob.glob(os.path.join(self.base_dir, '*.img')):
+            bbox = _ned_parse_filename(os.path.split(f)[1])
+            if bbox and self._intersects(bbox):
+                files.append(f)
+
+        args = ["gdalbuildvrt", "-q", self.vrt_file()] + files
+        status = subprocess.call(args)
```

Comment:
I couldn't find a Python interface to `gdalbuildvrt` itself, although there's functionality in the GDAL Python bindings to replicate its functionality. The question (#7) is how much there is in there that we'd need to reimplement. Perhaps there's an additional library for it?